equinox/nn/_composed.py

```diff
@@ -44,6 +44,8 @@ class MLP(Module):
     layers: Tuple[Linear, ...]
     activation: Callable
     final_activation: Callable
```

Comment:
@patrick-kidger i wasn't sure if these should have a `= static_field()` or not.  Tests pass, but I  don't have a mental model of what those are for.

Reply:
Hey there! Thanks for the PR.

So `static_field` is used to mark anything that should never interact with JAX transforms, like jit, grad etc. This is why it's on `in_size` for example -- this is just a bit of metadata, which we don't want to get promoted to a JAX array.

These deliberately don't have a `static_field`, as it is possible that someone may pass in a learnt activation function (e.g. `eqx.nn.PReLU`). In this case we do want JAX to be able to see whatever parameters are part of the activation function.

On which note -- you probably *do* want to put `static_field` on the extra bias values, which again are just extra metadata. (Other than that this PR LGTM.)

Reply:
I will add those then.

I think this is coming together in my head a bit. Thanks for your patience as I learn JAX. So these tag something so you can either put the equivalent to a `ZeroTangent`  (https://juliadiff.org/ChainRulesCore.jl/stable/api.html#ChainRulesCore.ZeroTangent ) either directly or by just filtering those out and stopping the recursion when building the tangent type? And hence callables like activation functions (or equinox modules) may themselves have nodes?

Reply:
A little bit, yes!

So the thing that's really going on here is what JAX calls "pytrees". These are any nested containers of dictionaries, tuples, lists, or user-registered custom nodes.

For example:
```python
pytree = [0, (1, 2)]
treedef = jax.tree_util.tree_structure(pytree)
print(treedef)  # PyTreeDef([*, (*, *)])
```
Likewise if you have no containers, then you just get a trivial pytree:
```python
pytree = 0
treedef = jax.tree_util.tree_structure(pytree)
print(treedef)  # PyTreeDef(*)
```
And in fact the types of the "leaves" of the tree is irrelevant to the structure:
```python
pytree = ["some string", object(), {"a": 0}]
treedef = jax.tree_util.tree_structure(pytree)
print(treedef)  # PyTreeDef([*, *, {"a": *}])
```

Every time you pass an object across any JAX API boundary, it will look over the leaves of the pytree structure -- e.g. when you do
```python
@jax.grad
def f(x):
    ...

x = (1.0, {"a": 2.0})
f(x)
```
then `x` is iterated over as a pytree, and gradients computed with respect to every leaf.

Alright, how does Equinox fit into this?

In Equinox, every module is *also* a pytree node. (One of those "user-registered custom nodes" from earlier.) For example `mlp = eqx.nn.MLP(...)` is a pytree -- this is what makes it possible to pass Equinox models across JAX's JIT/grad/etc. API boundaries, and get gradients computed wrt its leaves, for example.

And indeed, whether something is a pytree node and whether something is callable are unrelated concepts. For example this `mlp` is both a node and it is callable. (Contrast e.g. a dictionary, which is a node but not callable, or a regular function, which is callable but pytree-wise is "trivial" without subnodes.)

And finally, what does `static_field` do? It simply marks that this attribute is part of the structure of the pytree, and not a node:
```python
class Foo(eqx.Module):
    attribute_node: int
    attribute_structure: int = eqx.static_field()

foo = Foo(attribute_node=1, attribute_structure=2)
leaves, treedef = jax.tree_util.tree_flatten(foo)
print(leaves)  # [1]
print(treedef)  # PyTreeDef(CustomNode(Foo[('attribute_node',), ('attribute_structure',), (2,)], [*]))
```
Thus, it is "hidden" from any JAX transformations.

Reply:
One final addendum -- in an effort to provide a slightly more user-friendly API, Equinox provides some operations like `equinox.filter_{jit, grad, ...}` that are analogous to the original `jax.{jit, grad, ...}`. The difference is simply that Equinox automatically looks over all the leaves of the pytree, and only performs the transformation wrt the ones that are JAX or NumPy arrays. (And basically auto-static's everything else.)

As long as you're only using the `filter_*` verisons of transforms, this actually makes it possible to leave off `static_field` basically all of the time. (I still include the `static_field`s here just for the sake of anyone who happens not to be using these versions of the transforms.)

Tbh I think the fact that `jax.jit` does not already act like `equinox.filter_jit` (or `grad` like `filter_grad` etc.) was just a small mistake made early in JAX's API design, when things hadn't settled as much as they have now. In practice I usually just recommend that end users use `filter_{jit, grad, ...}` everywhere so that they don't need to worry about the details.